[#18] 회원 - 로그인, 로그아웃

sgo8308 opened the pull request:
### **개요**

로그인 로그아웃 기능을 추가했습니다.

### **이슈 링크**

#18 

### **작업 사항 (변경 사항)**
* 로그인하게 될 경우 세션에 LoginUser객체를 저장하게 됩니다.
* /login, /users를 제외한 모든 경로는 현재는 로그인을 하지 않으면 LoginCheckInterceptor에서 걸러집니다. 

### 테스트 내역 (o/x)
### **리뷰 받고 싶은 내용**
1. 
현재 여러 테스트들이 jsonPath를 이용하여 실제 응답 json의 값을 확인하고 있습니다. 예를 들어 패스워드를 입력하지 않았을 때 가입에 실패하는 테스트에서는 다음과 같은 json응답이 제대로 나가는지 확인합니다.
{
...
　"detail":[
　　　{
         　   　　　"field":"password"
        　　　}
    　　]
}

`mockMvc.perform(~~).andExpect(jsonPath("$.detail[0].field").value("password"))`


  
문제는 응답 json의 디테일한 사항에 너무 의존하고 있기 때문에 json 응답이 조금만 수정되도 테스트 코드가 계속 같이 수정되어야 하는 문제가 있었습니다. 실제로 개발하면서 에러메시지 json의 필드 값이 validationFieldError -> detail로 수정됨에 따라 테스트도 같이 수정 되어야 했습니다.

따라서 의존성을 덜기 위해 jsonPath검증을 빼고 싶은데요. 이렇게 된다면 이 테스트는 패스워드를 입력하지 않았을 때 BadRequest를 리턴하는지 아닌지만을 검증하게 됩니다. 실제로 패스워드를 입력하지 않은 것에 대해 실패를 일으킨게 아니라 다른 상황에 의해서 실패된 경우라면 이 테스트로는 제대로된 검증이 안되는 것입니다. 

이 문제에 대해서 멘토님의 의견을 듣고 싶습니다 !

2. 
Session과 관련된 상수들을 관리하기 위해 SessionConst라는 클래스를 만들었습니다.
문제는 이 클래스를 어떤 패키지에 넣어야 할지 잘 모르겠습니다. 일단 모든 서비스에서 공통으로 쓰일 수 있고 controller레이어에서 사용되기 때문에 common.controller 패키지에 넣었는데 더 좋은 방식은 없을지 궁금합니다. 그리고 패키지를 나누거나 어떤 클래스가 어떤 패키지에 들어갈지 정할 때 고려해야 할 사항들은 어떤 게 있을까요?

sgo8308 committed:
refactor(user): UserJoinController, UserService JoinController, JoinService로 이름 변경

- UserJoinController
User 패키지 아래 있고 JoinController 이름 만으로 무슨 컨트롤러인지 알 수 있다 판단하여 생략했습니다.

- UserService
서비스 클래스를 회원가입용, 로그인용으로 나누게 됨에 따라 더 구체적인 이름으로 변경했습니다.

sgo8308 committed:
refactor(user): 불필요하게 예외 던지는 코드 삭제

예외 반환 메소드에서 Excetion이 아닌 RuntimeException으로 반환함으로써 명시적으로 예외를 던지지 않아도 되게끔 구성했습니다.

sgo8308 committed:
refactor(user): JoinService의 identifyError() 안에 if문 switch로 가독성을 위해 변경

sgo8308 committed:
feat(user): login logout 기능 구현

필요한 값만 세션에 담는 것이 좋고 로그인 상태에서 사용자의 이름이 계속 필요하기 때문에 세션에 이름만 저장했습니다.

Closes #18

sgo8308 committed:
feat(user): 인터셉터를 활용하여 로그인 체크 구현

aop로 구현하면 권한 체크를 적용할 부분을 선택하기가 번거롭고 구현이 깔끔하지 않은 점
필터로 구현하면 예외 발생시 GlobalControllerAdvice에서 일관되게 처리할 수 없는 점
때문에 인터셉터를 활용했습니다.

Closes #18

sgo8308 committed:
feat(user): ArgumentResolver를 활용하여 컨트롤러에서 로그인 유저의 이름 바로 가져오기 구현

로그인 유저의 이름이 필요할 때마다 중복되는 작업을 ArgumentResolver를 활용하여 통일했습니다.

Closes #18

sgo8308 committed:
Merge branch f-lab-edu/develop into feature/18

sgo8308 committed:
chore(user): 불필요한 예외 던지는 코드 삭제

sgo8308 committed:
feat(user): user 정보 찾기 관련 API End Point 2개로 나누기

기존에는 아무 유저나 유저의 디테일한 정보에 접근 가능했던 문제가 있었음

sgo8308 committed:
feat(user):  LoginUser 클래스 추가, UserLoginDto LoginRequestDto로 변경

이전에는 유저의 이름만 저장햇으나 session에 저장했으나 user의 id도 저장하여 id로 유저를 찾을 수 있도록 추가

sgo8308 committed:
fix(user): 간헐적으로 로그인 테스트가 실패하는 문제 해결

- 문제 원인
PasswordEncryptionUtil에 toHex() 메소드에 16진수 값을 구한 후 왼쪽에 패딩 추가하는 로직이 없었음
그래서 같은 값이라도 toHex()하고 난 후 fromHex()하면 원본 값과 달라졌음

- 해결
toHex() 메소드에 패딩 추가하는 로직 추가

sgo8308 committed:
refactor(user): PasswordEncryptionUtil 매직 리터럴 제거

sgo8308 committed:
refactor(user): ShippingAddressMapper create메소드 이름 insert로 변경

mapper 레이어에서는 메소드 이름을 DB CRUD 이름과 비슷하게 가기로 결정

sgo8308 committed:
refactor(user): ShippingAddressMapper.xml create에 매핑된 이름 insert로 변경

sgo8308 committed:
test(user): LoginArgumentResolverTest 추가

sgo8308 committed:
chore(user): 패키지 이름 intercepter -> interceptor로 오타 수정

sgo8308 committed:
refactor(user): UserPrivateInfo에 정적 팩토리 메소드를 만들고 UserService에 적용

UserPrivateInfo를 생성하는 로직이 중복되고 깔끔하지 않아서 변경

derek-Flab commented on src/main/java/com/flab/marketgola/common/controller/SessionConst.java:
```diff
@@ -0,0 +1,12 @@
+package com.flab.marketgola.common.controller;
```
상수 클래스를 담는 패키지가 쫌 모호한거 같네요!?
controller에서만 쓰기 때문에 패키지를 담은거 같은데요. 

common패키지에 담은 이상 좀 더 범용적으로 사용하겠다는 뜻으로 해석되는게 보통이고, 
그렇기 때문에 
`constants` > `SessionConst(ant)`와 같이 패키지를 구성하는게 좀 더 좋을 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/argumentresolver/Login.java:
```diff
@@ -0,0 +1,12 @@
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Login {
```
이 어노테이션이 붙은 객체는 로그인된 유저 객체를 의미하는 것이지요?
Login이라는 어노테이션만 보니, 어디에 사용하는지 바로 이해가 되지 않는 것 같아서요. 
좀 더 명확하게 AuthenticatedUser가 어떨까요

derek-Flab commented on src/main/java/com/flab/marketgola/user/config/WebConfig.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void addArgumentResolvers(List<HandlerMethodArgumentResolver> resolvers) {
+        WebMvcConfigurer.super.addArgumentResolvers(resolvers);
```
이 설정코드는 꼭 넣어주어야 하나요?

derek-Flab commented on src/main/java/com/flab/marketgola/user/util/PasswordEncryptionUtil.java:
```diff
@@ -9,38 +12,81 @@
+            // 이전 iteration으로 암호화된 비밀번호를 알아내기 위하여 iteration과 digest를 같이 보관
+            return ENOUGH_ITERATION_COUNT + SEPERATION + toHex(salt) + SEPERATION + toHex(digest);
+        } catch (Exception e) {
+            throw new PasswordEncryptionFailException(e);
```
네이밍으로는 Password에 대한 정상적인 검증 실패로 보이는 네이밍인데, 실제 확인해보면 검증 하다가 내부적인 동작의 실패를 의미하는 exception이네요. 

패스워드 검증 실행에 문제가 생긴 네이밍으로 가져가도 좋을 것 같습니다. 

근데 패스워드를 검증하면서 내부적으로 코드가 실패되는 일은 버그 아닐까요?

이게 버그로 인해 문제가 발생하는 아주 예외적인 경우라면 try catch로 잡지 않고, 프로세스를 실행하면서 실패하면 throws Exception으로 던지고 internal server error를 찍도록 하는 것도 괜찮을 것 같습니다.

sgo8308 replied:
@derek-Flab 서버에서 발생한 예외는 최대한 숨기는게 좋다는 생각이 드는데요. 외부에서 내부 코드에 대해 직접적으로 알면 보안에 취약해질 것 같아서요 !

만약 위 코드에서 생기는 예외인 NoSuchAlgorithmException, InvalidKeySpecException 등을 그대로 던져버리면 errorCode로 NoSuchAlgorithm이나 InvalidKeySpec 같은 것이 나가게 됩니다. 저는 그것보다는 어떤 부분에서 생긴 예외인지 알 수 있게만 제가 직접 만든 예외(PasswordEncryptionFailException, PasswordValidationFailException)로 감싸서 내보내면서 Http status code는 500을 내보내면 내부 구조를 충분히 숨기면서도 서버에서 뭔가 문제가 발생했구나라는 것을 클라이언트 쪽에 인지시킬 수 있다고 판단했습니다.
그래서 throws Exception으로 직접 던지지 않고 try catch로 잡아서 예외를 감싸준 것인데요.

throws exception 방식으로 예외를 바로 던지면 어떤 점이 더 좋다고 할 수 있을까요 ~?

derek-Flab replied:
@sgo8308 
네네 서버에서 발생한 오류의 내용을 API의 메시지로 내보내지 않는게 맞습니다.
다른 예외처럼 로직상 발생 가능한 예외가 아니라, 
개발상 버그 라던지, 시스템적 문제로 발생하는 에러는 unExpectedError로 
{errorCode: "internal server error", status: 500, ...}이렇게 내려가게 되어있지 않나요? 
그렇게 처리해도 충분할 것 같다는 리뷰였습니다. 

설정 및 로직으로 인한 버그나 시스템적 오류는 저희가 대처하지 못한 것이기에 수정해야 할 목적이 있는 것이지? 그런 일이 발생 할 수 있기에 try catch로 감싸는 건 아닌 것 같습니다!

그리고 리뷰하다보니 
handleUnExpectedException예외인 경우 예상치 못한 오류로 인해 예외가 발생한 케이스인데, baseexception이 아닌 경우 로그 레벨이 debug여서 결국 prod(운영)환경에서는 어떤 로그도 발견하지 못할꺼 같은데 수정이 필요하지 않을까 싶네요? 

이해 안가는 부분 있다면 재질문 해주세요~

derek-Flab commented:
(1)
JSON응답이 조금 수정되어도 테스트 코드가 수정되는게 맞는 것 같습니다! 
이런 상황을 상상해보시죠. 어떤 response 필드를 개발해서 front end에서 사용하고 있습니다. 
근데 서버 개발자가 이 필드를 임의로 삭제해버리거나 이름을 바꿔버리게 되면 프론트엔드 코드는 바로 오류가 나버리게 되겠죠?
작업하면서 본의아니게 수정이 되었더라도, 이런걸 바로 잡아주기 위해서 테스트코드에 필드를 확인하는 작업이 있으면 리팩토링이 자유로울 것 입니다.



(2)
저는 여러 코드들을 보거나, 의논하면서 패키지를 정해온 경험을 바탕으로 하고 있어서 특별히 법칙이 있는 것은 아닌 것 같습니다. 
좀 괜찮다 싶은 오픈소스를 보면서 참고 해보는것도 좋을 것 같습니다. 
저같은 경우는 common이라는 패키지에서 상수 클래스를 관리한다면 constants라는 패키지에 목적에 맞는 contant클래스 이름을 지어서 활용하고 있습니다. 
좀 한가지 체크리스트로 가져갈만한 부분은요. 
1. constant가 해당 user패키지에서만 사용할 것 같다면? user package내부에 선언해주는게 좋습니다. 근데 여러 곳에서 사용할 것 같다? 예를 들면 `TimeUtil`(날짜, 시간을 처리해주는 유틸. 딱 봐도 범용적인?)이런것은 글로벌 패키지에 두겠죠?
2. ~~XXX.java일 때 XXX를 패키지명으로 해도 괜찮을 것 같아요(만능 키는 아닙니다). xxxUtil -> util패키지, xxxConstant -> constant패키지, xxxController -> controller 패키지 이런식으로요. 간단하지만, 다른 사람들이 읽을 때 쉽게 유추해서 찾아들어갈 수 있을 것 같습니다.

derek-Flab commented:
수고하셨습니다!ㅎㅎ코드 해석하는데 문제 없었고, 슬랙으로 말씀하신 session이 controller, service어디에 위치해야하는지에 대한 고민도 어느정도 이해가 갑니다. 세션은 현재 작업하신 것처럼 controller 레이어에서 컨트롤하는게 좋을 것 같다는 의견입니다.

sgo8308 committed:
refactor(user): @Login 어노테이션 명확하게 @AuthenticatedUser로 변경

sgo8308 committed:
refactor(common): SessionConst -> SessionConstant로 변경, 패키지는 constant 패키지로 옮김

클래스 이름을 일관되게 줄여쓰지 않고 표현하고 패키지는 좀 더 쉽게 찾을 수 있게 하기 위하여 변경

sgo8308 committed:
refactor(user): WebConfig.addArgumentResolvers()에서 불필요한 설정코드 제거

sgo8308 committed:
refactor(user): PasswordEncryptionUtil에서 예외들을 그대로 밖으로 던지도록 변경

RuntimeException으로 감싸서 던질 수도 있지만 실제로 RuntimeException이 아닌데,
이렇게 감싸 던지는 경우 실제 예외가 발생했을 때 다른 개발자들에게 혼동을 줄 여지가 있다고 판단

sgo8308 committed:
chore(user): 필요없어진 PasswordEncryptionFailException, PasswordValidationFailException 제거

sgo8308 committed:
refactor(common): 예상치 못한 예외들은 GlobalControllerAdivce에서 UnExceptedException으로 변환되서 처리되도록 변경

프로그램 내부에서 발생한 문제는 보안상 외부로 노출하지 않고 errorCode: UnExpected로 통일해서 나갈 수 있게 하기 위하여
